File: src/scoreboard/ScoreboardManager.ts
import { Player, Server } from "../types/hytale";
import { ScoreboardConfig } from "../config/scoreboard.config";
import { PlaceholderRegistry } from "./PlaceholderRegistry";
import { PlayerDataProvider } from "../data/PlayerDataProvider";
import { TestDataProvider } from "../data/TestDataProvider";
import { ScoreboardSession } from "./ScoreboardSession";

export class ScoreboardManager {
  private readonly server: Server;
  private readonly config: ScoreboardConfig;
  private readonly placeholderRegistry: PlaceholderRegistry;
  private readonly dataProvider: PlayerDataProvider;
  private readonly testDataProvider: TestDataProvider;
  private readonly sessions = new Map<string, ScoreboardSession>();
  private updateTask: { cancel: () => void } | null = null;

  constructor(
    server: Server,
    config: ScoreboardConfig,
    placeholderRegistry: PlaceholderRegistry,
    dataProvider: PlayerDataProvider,
    testDataProvider: TestDataProvider,
  ) {
    this.server = server;
    this.config = config;
    this.placeholderRegistry = placeholderRegistry;
    this.dataProvider = dataProvider;
    this.testDataProvider = testDataProvider;
  }

  start(): void {
    this.server.events.onPlayerJoin((player) => this.createSession(player));
    this.server.events.onPlayerLeave((player) => this.removeSession(player));

    this.updateTask = this.server.scheduler.runRepeatingTask(
      () => this.updateAll(),
      this.config.updateIntervalMs,
    );
  }

  stop(): void {
    this.updateTask?.cancel();
    this.updateTask = null;

    for (const session of this.sessions.values()) {
      session.destroy();
    }
    this.sessions.clear();
  }

  setEnabled(player: Player, enabled: boolean): void {
    const session = this.sessions.get(player.id);
    if (!session) {
      return;
    }

    session.setEnabled(enabled);
  }

  setTestMode(player: Player, enabled: boolean): void {
    const session = this.sessions.get(player.id);
    if (!session) {
      return;
    }

    session.setTestMode(enabled);
  }

  private createSession(player: Player): void {
    const sidebar = this.server.createScoreboardSidebar(player);
    const session = new ScoreboardSession(player, sidebar, this.config.defaultEnabled);
    this.sessions.set(player.id, session);

    if (session.isEnabled()) {
      sidebar.show();
    }
  }

  private removeSession(player: Player): void {
    const session = this.sessions.get(player.id);
    if (!session) {
      return;
    }

    session.destroy();
    this.sessions.delete(player.id);
  }

  private updateAll(): void {
    const onlineStatus = this.server.getOnlineStatus();

    for (const session of this.sessions.values()) {
      if (!session.isEnabled()) {
        continue;
      }

      const dataProvider = session.isTestMode() ? this.testDataProvider : this.dataProvider;
      const playerData = dataProvider.getPlayerData(session.player);

      const context = {
        player: session.player,
        online: onlineStatus.online,
        max: onlineStatus.max,
        gameName: playerData.gameName,
        data: {
          rank: playerData.rank,
          coins: playerData.coins,
          level: playerData.level,
        },
      };

      const title = this.placeholderRegistry.resolveLine(this.config.title, context);
      const lines = this.config.lines.map((line) =>
        this.placeholderRegistry.resolveLine(line, context),
      );

      session.update(title, this.dedupeLines(lines));
    }
  }

  private dedupeLines(lines: string[]): string[] {
    const seen = new Map<string, number>();
    return lines.map((line) => {
      const count = seen.get(line) ?? 0;
      seen.set(line, count + 1);
      if (count === 0) {
        return line;
      }
      return `${line}\u200B${count}`;
    });
  }
}

File: src/main/java/com/hytale/scoreboard/scoreboard/ScoreboardManager.java
package com.hytale.scoreboard.scoreboard;

import com.hytale.scoreboard.config.ScoreboardConfig;
import com.hytale.scoreboard.data.PlayerData;
import com.hytale.scoreboard.data.PlayerDataProvider;
import com.hytale.scoreboard.data.TestDataProvider;
import com.hytale.scoreboard.types.OnlineStatus;
import com.hytale.scoreboard.types.Player;
import com.hytale.scoreboard.types.ScheduledTask;
import com.hytale.scoreboard.types.Server;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ScoreboardManager {
  private final Server server;
  private final ScoreboardConfig config;
  private final PlaceholderRegistry placeholderRegistry;
  private final PlayerDataProvider dataProvider;
  private final TestDataProvider testDataProvider;
  private final Map<String, ScoreboardSession> sessions = new HashMap<>();
  private ScheduledTask updateTask;

  public ScoreboardManager(
      Server server,
      ScoreboardConfig config,
      PlaceholderRegistry placeholderRegistry,
      PlayerDataProvider dataProvider,
      TestDataProvider testDataProvider) {
    this.server = server;
    this.config = config;
    this.placeholderRegistry = placeholderRegistry;
    this.dataProvider = dataProvider;
    this.testDataProvider = testDataProvider;
  }

  public void start() {
    server.getEvents().onPlayerJoin(this::createSession);
    server.getEvents().onPlayerLeave(this::removeSession);

    updateTask = server.getScheduler().runRepeatingTask(this::updateAll, config.getUpdateIntervalMs());
  }

  public void stop() {
    if (updateTask != null) {
      updateTask.cancel();
      updateTask = null;
    }

    for (ScoreboardSession session : sessions.values()) {
      session.destroy();
    }
    sessions.clear();
  }

  public void setEnabled(Player player, boolean enabled) {
    ScoreboardSession session = sessions.get(player.getId());
    if (session == null) {
      return;
    }

    session.setEnabled(enabled);
  }

  public void setTestMode(Player player, boolean enabled) {
    ScoreboardSession session = sessions.get(player.getId());
    if (session == null) {
      return;
    }

    session.setTestMode(enabled);
  }

  private void createSession(Player player) {
    var sidebar = server.createScoreboardSidebar(player);
    ScoreboardSession session = new ScoreboardSession(player, sidebar, config.isDefaultEnabled());
    sessions.put(player.getId(), session);

    if (session.isEnabled()) {
      sidebar.show();
    }
  }

  private void removeSession(Player player) {
    ScoreboardSession session = sessions.get(player.getId());
    if (session == null) {
      return;
    }

    session.destroy();
    sessions.remove(player.getId());
  }

  private void updateAll() {
    OnlineStatus onlineStatus = server.getOnlineStatus();

    for (ScoreboardSession session : sessions.values()) {
      if (!session.isEnabled()) {
        continue;
      }

      PlayerDataProvider provider = session.isTestMode() ? testDataProvider : dataProvider;
      PlayerData playerData = provider.getPlayerData(session.getPlayer());

      Map<String, Object> data = new HashMap<>();
      data.put("rank", playerData.getRank());
      data.put("coins", playerData.getCoins());
      data.put("level", playerData.getLevel());

      PlaceholderRegistry.PlaceholderContext context = new PlaceholderRegistry.PlaceholderContext(
          session.getPlayer(),
          onlineStatus.getOnline(),
          onlineStatus.getMax(),
          playerData.getGameName(),
          data);

      String title = placeholderRegistry.resolveLine(config.getTitle(), context);
      List<String> lines = new ArrayList<>();
      for (String line : config.getLines()) {
        lines.add(placeholderRegistry.resolveLine(line, context));
      }

      session.update(title, dedupeLines(lines));
    }
  }

  private List<String> dedupeLines(List<String> lines) {
    Map<String, Integer> seen = new HashMap<>();
    List<String> deduped = new ArrayList<>();
    for (String line : lines) {
      int count = seen.getOrDefault(line, 0);
      seen.put(line, count + 1);
      if (count == 0) {
        deduped.add(line);
      } else {
        deduped.add(line + "\u200B" + count);
      }
    }
    return deduped;
  }
}
